Loki 2.4 crashes every 10 minutes

<!-- Please keep the structure below, or your issue may be closed. -->

**Describe the bug**
After upgrading loki to v2.4 it crashes every 10 minutes:
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x1af803c]

goroutine 9347 [running]:
github.com/grafana/loki/pkg/validation.(*Overrides).AllByUserID(0x0)
        /src/loki/pkg/validation/limits.go:254 +0x1c
github.com/grafana/loki/pkg/storage/stores/shipper/compactor/retention.findSmallestRetentionPeriod({0x26f81e8, 0xc00054bf20})
        /src/loki/pkg/storage/stores/shipper/compactor/retention/expiration.go:126 +0x82
github.com/grafana/loki/pkg/storage/stores/shipper/compactor/retention.(*expirationChecker).MarkPhaseStarted(0xc000711c50)
        /src/loki/pkg/storage/stores/shipper/compactor/retention/expiration.go:60 +0x45
github.com/grafana/loki/pkg/storage/stores/shipper/compactor.(*expirationChecker).MarkPhaseStarted(0xc0000a71a0)
        /src/loki/pkg/storage/stores/shipper/compactor/compactor.go:516 +0x29
github.com/grafana/loki/pkg/storage/stores/shipper/compactor.(*Compactor).RunCompaction(0xc000222000, {0x26f7500, 0xc0011f7840})
        /src/loki/pkg/storage/stores/shipper/compactor/compactor.go:409 +0xb6
github.com/grafana/loki/pkg/storage/stores/shipper/compactor.(*Compactor).runCompactions.func1()
        /src/loki/pkg/storage/stores/shipper/compactor/compactor.go:342 +0x29
github.com/grafana/loki/pkg/storage/stores/shipper/compactor.(*Compactor).runCompactions.func2()
        /src/loki/pkg/storage/stores/shipper/compactor/compactor.go:350 +0x9e
created by github.com/grafana/loki/pkg/storage/stores/shipper/compactor.(*Compactor).runCompactions
        /src/loki/pkg/storage/stores/shipper/compactor/compactor.go:348 +0x434
```

**Environment:**
 - Infrastructure: docker swarm
 - Deployment tool: docker swarm

**Screenshots, Promtail config, or terminal output**
My config:
```
auth_enabled: false

server:
  http_listen_port: 3100

common:
  path_prefix: /loki
  storage:
    filesystem:
      chunks_directory: /loki/chunks
      rules_directory: /loki/rules
  replication_factor: 1
  ring:
    instance_addr: 127.0.0.1
    kvstore:
      store: inmemory

schema_config:
  configs:
  - from: 2018-04-15
    store: boltdb
    object_store: filesystem
    schema: v11
    index:
      prefix: index_
      period: 168h
  - from: 2022-11-07
    store: boltdb-shipper
    object_store: filesystem
    schema: v11
    index:
      prefix: index2_
      period: 24h

storage_config:
  boltdb:
    directory: /loki/index
  filesystem:
    directory: /loki/chunks

compactor:
  retention_enabled: true

limits_config:
  retention_period: 90d
```

Also having the same issue on loki 2.4 version. Running "Compactor" component as standalone service. It crashes every 10 min. with following error.

```level=info ts=2021-11-07T11:29:37.546535802Z caller=compactor.go:337 msg="compactor startup delay completed"
level=info ts=2021-11-07T11:29:37.546616775Z caller=compactor.go:376 msg="compactor started"
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x1af803c]

goroutine 15902 [running]:
github.com/grafana/loki/pkg/validation.(*Overrides).AllByUserID(0xc0005d8980)
	/src/loki/pkg/validation/limits.go:254 +0x1c
github.com/grafana/loki/pkg/storage/stores/shipper/compactor/retention.findSmallestRetentionPeriod({0x26f81e8, 0xc00000c9a8})
	/src/loki/pkg/storage/stores/shipper/compactor/retention/expiration.go:126 +0x82
github.com/grafana/loki/pkg/storage/stores/shipper/compactor/retention.(*expirationChecker).MarkPhaseStarted(0xc0004fdbb0)
	/src/loki/pkg/storage/stores/shipper/compactor/retention/expiration.go:60 +0x45
github.com/grafana/loki/pkg/storage/stores/shipper/compactor.(*expirationChecker).MarkPhaseStarted(0xc0005cd120)
	/src/loki/pkg/storage/stores/shipper/compactor/compactor.go:516 +0x29
github.com/grafana/loki/pkg/storage/stores/shipper/compactor.(*Compactor).RunCompaction(0xc0000e9b00, {0x26f7500, 0xc000522080})
	/src/loki/pkg/storage/stores/shipper/compactor/compactor.go:409 +0xb6
github.com/grafana/loki/pkg/storage/stores/shipper/compactor.(*Compactor).runCompactions.func1()
	/src/loki/pkg/storage/stores/shipper/compactor/compactor.go:342 +0x29
github.com/grafana/loki/pkg/storage/stores/shipper/compactor.(*Compactor).runCompactions.func2()
	/src/loki/pkg/storage/stores/shipper/compactor/compactor.go:350 +0x9e
created by github.com/grafana/loki/pkg/storage/stores/shipper/compactor.(*Compactor).runCompactions
	/src/loki/pkg/storage/stores/shipper/compactor/compactor.go:348 +0x434

--- a/pkg/validation/limits.go
+++ b/pkg/validation/limits.go
@@ -251,7 +251,12 @@ func NewOverrides(defaults Limits, tenantLimits TenantLimits) (*Overrides, error
 	}, nil
 }
 
-func (o *Overrides) AllByUserID() map[string]*Limits { return o.tenantLimits.AllByUserID() }
+func (o *Overrides) AllByUserID() map[string]*Limits {
+	if o.tenantLimits != nil {
+		return o.tenantLimits.AllByUserID()
+	}
+	return nil
+}
 
 // IngestionRateStrategy returns whether the ingestion rate limit should be individually applied
 // to each distributor instance (local) or evenly shared across the cluster (global).

